Add MF2-ish message/resource representations

eemeli opened the pull request:
Adds
```py
from compare_locales.resource import resourceFromFluent, resourceFromProperties
```
along with data class definitions for a uniform representation of message resources as
```py
list[Comment | Message | Junk]
```

Constructors for other formats may be added later, as well as common linters and validators.

eviljeff commented:
Is there an issue/ticket for this change?  (I don't know what MF2 is!)

eemeli commented:
@eviljeff Sorry for the delay in answering, I'd managed to miss your question. 🙈

MF2 is MessageFormat 2, and in this context may be thought of as a family of specifications that provide a successor to Fluent. The [single-message spec](https://github.com/unicode-org/message-format-wg/tree/main/spec) is pretty well developed, and the message data model that's introduced here is close to its data model.

A key reason for using this data model is that MF2 is a superset of all the features in all the formats we care about, and its data model can therefore represent all the messages we care about independently of their source format. This will lets us build validators, linters, and transformers for messages separately from their syntax parsers and serialisers.

eviljeff commented on compare_locales/resource/from_properties.py:
```diff
@@ -0,0 +1,68 @@
+            if start > prevEnd:
+                pattern.append(Text(value[prevEnd:start]))
+            pattern.append(VariableRef(match[0]))
+            prevEnd = match.end()
```
could this result in `prevEnd` not always increasing?  I.e. should it be `prevEnd = max(prevEnd, match.end())`?

eemeli replied:
Because `.finditer()` only returns non-overlapping matches left-to-right, using `max()` would not change the behaviour of this line. The only way for `match.end()` to not be greater than `prevEnd` is if the match is empty. Which is technically valid, but mostly nonsensical.

eviljeff commented on compare_locales/resource/from_properties.py:
```diff
@@ -0,0 +1,68 @@
+    """
+
+    pattern: Pattern = []
+    prevEnd = 0
```
camelCase

eviljeff commented on compare_locales/resource/from_properties.py:
```diff
@@ -0,0 +1,68 @@
+                pattern.append(Text(value[prevEnd:start]))
+            pattern.append(VariableRef(match[0]))
+            prevEnd = match.end()
+    if len(pattern) == 0:
+        pattern.append(Text(value))
```
I think this clause is redundant because `prevEnd` will be `0` if `len(pattern) == 0` so the `elif` would match&work for it too.

eemeli commented on compare_locales/resource/from_fluent.py:
```diff
@@ -0,0 +1,255 @@
+    def add(arg: SelectArg) -> None:
+        prev = next((a for a in args if a.selector.equals(arg.selector)), None)
+        if prev:
+            prev.keys += arg.keys
```
It isn't lost; thanks to the `.equals()` it needs to be the same for `prev` and `arg`. In the rare case that we've two Fluent selectors within a single message that otherwise match but use a different name for their default variant, we end up considering those selectors as separate in the generated data model so that the `defaultName` isn't lost.

But if that happens, we really ought to complain about it when linting as it's probably a mistake in the source message.

eemeli commented on compare_locales/resource/from_fluent.py:
```diff
@@ -0,0 +1,255 @@
+            options[arg.name.name] = (
+                Literal(False, arg.value.value)
+                if isinstance(arg.value, Fluent.NumberLiteral)
+                else Literal(True, arg.value.parse()["value"])
```
Yes, it always has that shape: https://github.com/projectfluent/python-fluent/blob/fd57cd024e0c2fae4e33014f773e1500a5d34417/fluent.syntax/fluent/syntax/ast.py#L205-L206

eemeli commented on compare_locales/resource/from_fluent.py:
```diff
@@ -0,0 +1,255 @@
+        if i == 0:
+            keys = [[key] for key in kk]
+        else:
+            for i in range(len(keys), 0, -1):
+                prev = keys[i - 1]
+                keys[i - 1 : i] = [prev + [key] for key in kk]
```
This clearly needs a bit more explanation, and not just in this PR.

Do this and the above code comment clarify what's happening here?

```suggestion
            # For selectors after the first,
            # replace each previous variant with len(kk) variants,
            # one for each of this selector's keys.
            for j in range(len(keys), 0, -1):
                prev = keys[j - 1]
                keys[j - 1 : j] = [prev + [key] for key in kk]
```

eemeli commented on compare_locales/resource/from_fluent.py:
```diff
@@ -0,0 +1,255 @@
+    if len(args) == 0:
+        return PatternMessage([elementToPart(el) for el in ast.elements])
+
+    # First determine the keys for all cases, with empty values
```
```suggestion
    # First determine the keys for all variants, with empty values.
    # Fluent supports selectors within a message,
    # while our data model supports top-level selectors only.
    # Mapping between these approaches gets a bit complicated
    # when a message contains more than one selector.
```

eemeli commented on compare_locales/resource/from_properties.py:
```diff
@@ -0,0 +1,68 @@
+    """
+
+    pattern: Pattern = []
+    prevEnd = 0
+    if variables:
+        for match in variables.finditer(value):
+            start = match.start()
+            if start > prevEnd:
+                pattern.append(Text(value[prevEnd:start]))
+            pattern.append(VariableRef(match[0]))
+            prevEnd = match.end()
+    if len(pattern) == 0:
+        pattern.append(Text(value))
+    elif prevEnd < len(value):
+        pattern.append(Text(value[prevEnd:]))
```
```suggestion
    prev_end = 0
    if variables:
        for match in variables.finditer(value):
            start = match.start()
            if start > prev_end:
                pattern.append(Text(value[prev_end:start]))
            pattern.append(VariableRef(match[0]))
            prev_end = match.end()
    if len(pattern) == 0 or prev_end < len(value):
        pattern.append(Text(value[prev_end:]))
```

eemeli commented:
@eviljeff Thanks to your prodding in https://github.com/mozilla/compare-locales/pull/19#discussion_r1276634456 I went ahead and added a [test case](https://github.com/mozilla/compare-locales/pull/19/files#diff-f6ca6370655fb06ce7e72261559dabc290ea6a306e57507b4af94abe2da49627R180) that exercises a part of the code that wasn’t well covered before, and it turned out to have a small bug that I now also fixed.

With that, we start from a Fluent message
```fluent
abc =
  { $a ->
     [one]
       { $b ->
          [two] one-two
         *[other] one-other
       }
    *[other]
       { $b ->
          [two] other-two
         *[other] other-other
       }
  }
```
and we need to end up with a structure that corresponds to this in MF2 syntax:
```
match {$a} {$b}
when one two {one-two}
when one * {one-other}
when * two {other-two}
when * * {other-other}
```

To achieve that, in `messageFromFluentPattern()` first the `args` list of selectors is found, with references to the two unique selectors `$a` and `$b`. Then `keys` is determined by iterating through those.

On the first pass, the keys `kk` for `$a` are found to be `['one', None]` (with `None` indicating the catchall variant), so `keys` is set to `[['one'], [None]]`.

On the second pass, the keys `kk` for `$b` are found to be `['two', None]` (they're the same in both places in Fluent), and we need to extend `keys` to generate the four variants that match the MF2 structure. So, iterating from the end of `keys` to make the indexing easier we have:
1. Start with `keys = [['one'], [None]]` and `kk = ['two', None]`.
2. Replace the `[None]` with two entries `[None, 'two'], [None, None]`.
3. `keys` is now `[['one'], [None, 'two'], [None, None]]`.
4. Replace the `['one']` with two entries `['one', 'two'], ['one', None]`.
5. End up with `keys = [['one', 'two'], ['one', None], [None, 'two'], [None, None]]`.

From that, we can build the `variants` list initially with empty pattern values, and then work our way through the Fluent message again to fill those out. There's probably a way to do all this with a single pass in most cases, but this works relatively efficiently and doesn't require any backtracking if you discover late that e.g. two inner selectors aren't quite equivalent after all.
